File: ai/speed/detect.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cmath>
#include <tuple>
#include <algorithm>

class BallDetector {
public:
    BallDetector(bool showDisplay = false, bool recordVideo = false, 
                 double focalLength = 554.26, double realDiameter = 0.04)
      : showDisplay_(showDisplay), recordVideo_(recordVideo),
        focalLength_(focalLength), realDiameter_(realDiameter)
    {
        // 打开摄像头
        cap_.open(0);
        if (!cap_.isOpened()) {
            std::cerr << "无法打开摄像头!\n";
            exit(-1);
        }
        cap_.set(cv::CAP_PROP_FRAME_WIDTH, 640);
        cap_.set(cv::CAP_PROP_FRAME_HEIGHT, 480);
        cap_.set(cv::CAP_PROP_FPS, 30);

        // 默认标定参数（可通过 FileStorage 加载 calib 文件）
        cameraMatrix_ = (cv::Mat_<double>(3,3) << focalLength_, 0, 320,
                                                  0, focalLength_, 240,
                                                  0, 0, 1);
        distortionCoeffs_ = cv::Mat::zeros(1, 5, CV_64F);

        // 参数设定（这里采用硬编码阈值，可根据实际情况调整）
        orangeLower_ = cv::Scalar(5, 100, 100);
        orangeUpper_ = cv::Scalar(15, 255, 255);

        whiteLower_ = 200;
        whiteUpper_ = 255;
        whiteBlur_ = 7;

        orangeIterOpen_ = 1;
        orangeIterClose_ = 1;
        whiteIterOpen_ = 1;
        whiteIterClose_ = 1;
        minRadius_ = 5;
        maxRadius_ = 50;
        minCircularity_ = 0.7;
        glareThresh_ = 200;

        kernel_ = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5,5));

        frameCount_ = 0;
        startTime_ = cv::getTickCount();

        if(recordVideo_) {
            videoWriter_.open("C:/usr/src/ai/record/detection.avi", 
                cv::VideoWriter::fourcc('X','V','I','D'), 8, cv::Size(640,480));
            if (!videoWriter_.isOpened()) {
                std::cerr << "无法打开视频写入器!\n";
            }
        }
    }
    
    void run() {
        cv::Mat frame, processedFrame;
        while (true) {
            if (!cap_.read(frame)) break;

            int status, x, y;
            double distance, angle;
            std::tie(status, x, y, distance, angle, processedFrame) = detectBalls(frame);
            
            frameCount_++;
            double elapsedTime = (cv::getTickCount() - startTime_) / cv::getTickFrequency();
            double fps = frameCount_ / elapsedTime;
            
            if (status == 1) {
                std::cout << "位置: (" << x << ", " << y << "), 距离: " 
                          << distance << "米, 角度: " << angle << "度, FPS: " 
                          << fps << "\n";
            }
            
            if (showDisplay_) {
                cv::imshow("Frame", processedFrame);
                if (cv::waitKey(1) == 'q') break;
            }
            
            if(recordVideo_ && videoWriter_.isOpened()) {
                videoWriter_.write(processedFrame);
            }
        }
        
        cap_.release();
        if(videoWriter_.isOpened())
            videoWriter_.release();
        cv::destroyAllWindows();
    }
    
private:
    // 返回值依次为 status, x, y, distance, angle, 以及带调试信息的帧
    std::tuple<int, int, int, double, double, cv::Mat> detectBalls(const cv::Mat &frame) {
        cv::Mat undistorted;
        cv::undistort(frame, undistorted, cameraMatrix_, distortionCoeffs_);
        
        // 复用灰度图计算 HSV 中的 V 通道
        cv::Mat gray;
        cv::cvtColor(undistorted, gray, cv::COLOR_BGR2GRAY);
        
        cv::Mat hsv;
        cv::cvtColor(undistorted, hsv, cv::COLOR_BGR2HSV);
        std::vector<cv::Mat> hsvChannels;
        cv::split(hsv, hsvChannels);
        hsvChannels[2] = gray;
        cv::merge(hsvChannels, hsv);
        cv::Mat frameModified;
        cv::cvtColor(hsv, frameModified, cv::COLOR_HSV2BGR);
        
        cv::Mat glareFree = glareSuppression(frameModified);
        
        // 分别处理橙色和白色小球（若需要可并行化）
        cv::Mat orangeMask = processOrange(glareFree);
        cv::Mat whiteMask = processWhite(glareFree);
        
        std::vector<cv::Vec3f> orangeBalls = findTarget(orangeMask);
        std::vector<cv::Vec3f> whiteBalls = findTarget(whiteMask);
        
        if (!orangeBalls.empty()) {
            return calculateMetrics(orangeBalls[0], frame);
        } else if (!whiteBalls.empty()) {
            return calculateMetrics(whiteBalls[0], frame);
        } else {
            return std::make_tuple(0, 0, 0, 0.0, 0.0, frame);
        }
    }
    
    cv::Mat processOrange(const cv::Mat &frame) {
        cv::Mat hsv, mask;
        cv::cvtColor(frame, hsv, cv::COLOR_BGR2HSV);
        cv::inRange(hsv, orangeLower_, orangeUpper_, mask);
        cv::morphologyEx(mask, mask, cv::MORPH_OPEN, kernel_, cv::Point(-1,-1), orangeIterOpen_);
        cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, kernel_, cv::Point(-1,-1), orangeIterClose_);
        return mask;
    }
    
    cv::Mat processWhite(const cv::Mat &frame) {
        cv::Mat gray, blurred, mask;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        cv::GaussianBlur(gray, blurred, cv::Size(whiteBlur_, whiteBlur_), 0);
        cv::threshold(blurred, mask, whiteLower_, whiteUpper_, cv::THRESH_BINARY);
        cv::morphologyEx(mask, mask, cv::MORPH_OPEN, kernel_, cv::Point(-1,-1), whiteIterOpen_);
        cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, kernel_, cv::Point(-1,-1), whiteIterClose_);
        return mask;
    }
    
    // 返回 Vec3f：x, y, 直径（2 * radius）
    std::vector<cv::Vec3f> findTarget(const cv::Mat &mask) {
        std::vector<std::vector<cv::Point>> contours;
        cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
        std::vector<cv::Vec3f> valid;
        for (const auto &cnt : contours) {
            double area = cv::contourArea(cnt);
            if (area < 100)
                continue;
            double perimeter = cv::arcLength(cnt, true);
            if (perimeter == 0)
                continue;
            double circularity = (4 * CV_PI * area) / (perimeter * perimeter);
            cv::Point2f center;
            float radius;
            cv::minEnclosingCircle(cnt, center, radius);
            if (radius >= minRadius_ && radius <= maxRadius_ && circularity >= minCircularity_) {
                valid.push_back(cv::Vec3f(center.x, center.y, radius * 2));
            }
        }
        return valid;
    }
    
    std::tuple<int, int, int, double, double, cv::Mat> calculateMetrics(const cv::Vec3f &ballInfo, const cv::Mat &frame) {
        int x = static_cast<int>(ballInfo[0]);
        int y = static_cast<int>(ballInfo[1]);
        float diameter = ballInfo[2];
        if(diameter <= 0 || x < 0 || x >= 640 || y < 0 || y >= 480)
            return std::make_tuple(0, 0, 0, 0.0, 0.0, frame);
        double distance = (realDiameter_ * focalLength_) / std::max(diameter, 1e-6f);
        int dx = x - 320;
        double angle = focalLength_ != 0 ? std::atan(dx / focalLength_) * 180.0 / CV_PI : 0;
        cv::Mat debugFrame = frame.clone();
        if(showDisplay_ || recordVideo_)
            drawDebugInfo(debugFrame, x, y, distance, angle);
        return std::make_tuple(1, x, y, distance, angle, debugFrame);
    }
    
    cv::Mat glareSuppression(const cv::Mat &frame) {
        cv::Mat gray;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        double avgBrightness = cv::mean(gray)[0];
        if (avgBrightness < 100)
            return frame;
        cv::Mat glareMask, result;
        cv::threshold(gray, glareMask, glareThresh_, 255, cv::THRESH_BINARY);
        cv::inpaint(frame, glareMask, result, 3, cv::INPAINT_TELEA);
        return result;
    }
    
    void drawDebugInfo(cv::Mat &frame, int x, int y, double distance, double angle) {
        cv::line(frame, cv::Point(320, 0), cv::Point(320, 480), cv::Scalar(100,100,100), 1);
        cv::line(frame, cv::Point(0, 240), cv::Point(640, 240), cv::Scalar(100,100,100), 1);
        cv::circle(frame, cv::Point(x,y), 10, cv::Scalar(0,255,0), 2);
        cv::line(frame, cv::Point(320,240), cv::Point(x,y), cv::Scalar(0,0,255), 2);
        cv::putText(frame, cv::format("Dist: %.2f m", distance), cv::Point(10,30),
                    cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0,255,255), 2);
        cv::putText(frame, cv::format("Angle: %.1f deg", angle), cv::Point(10,70),
                    cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0,255,255), 2);
    }
    
    cv::VideoCapture cap_;
    cv::VideoWriter videoWriter_;
    bool showDisplay_;
    bool recordVideo_;
    double focalLength_, realDiameter_;
    cv::Mat cameraMatrix_, distortionCoeffs_;
    cv::Scalar orangeLower_, orangeUpper_;
    int whiteLower_, whiteUpper_;
    int whiteBlur_;
    int orangeIterOpen_, orangeIterClose_;
    int whiteIterOpen_, whiteIterClose_;
    int minRadius_, maxRadius_;
    double minCircularity_;
    int glareThresh_;
    cv::Mat kernel_;
    
    int frameCount_;
    int64 startTime_;
};

int main() {
    // 设为 true 可显示检测窗口及打印调试信息
    BallDetector detector(true, true);
    detector.run();
    return 0;
}

File: ai/detect.py
import cv2
import numpy as np
import time
import math
import configparser

class BallDetector:
    def __init__(self, show_display=False, record_video=False, focal_length=275, real_diameter=0.04, config_file='/usr/src/ai/config.txt'):
        config = configparser.ConfigParser()
        config.read(config_file)
        
        self.cap = cv2.VideoCapture(0)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)
        self.cap.set(cv2.CAP_PROP_FPS, 30)
        self.cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*'MJPG'))
        
        self.width = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        self.height = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        print(f"设置后的分辨率: {self.width}x{self.height}")
        
        self.focal_length = focal_length
        self.real_diameter = real_diameter
        
        self.show_display = show_display
        self.record_video = record_video
        self.kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (5, 5))
        
        self.orange_lower = np.array([config.getint('DEFAULT', 'orange_lower_h'),
                                      config.getint('DEFAULT', 'orange_lower_s'),
                                      config.getint('DEFAULT', 'orange_lower_v')])
        self.orange_upper = np.array([config.getint('DEFAULT', 'orange_upper_h'),
                                      config.getint('DEFAULT', 'orange_upper_s'),
                                      config.getint('DEFAULT', 'orange_upper_v')])
        
        self.white_lower = config.getint('DEFAULT', 'white_lower')
        self.white_upper = config.getint('DEFAULT', 'white_upper')
        self.white_blur = config.getint('DEFAULT', 'white_blur')
        
        self.orange_iter_open = config.getint('DEFAULT', 'orange_iter_open')
        self.orange_iter_close = config.getint('DEFAULT', 'orange_iter_close')
        self.white_iter_open = config.getint('DEFAULT', 'white_iter_open')
        self.white_iter_close = config.getint('DEFAULT', 'white_iter_close')
        
        self.min_radius = config.getint('DEFAULT', 'min_radius')
        self.max_radius = config.getint('DEFAULT', 'max_radius')
        self.min_circularity = config.getfloat('DEFAULT', 'min_circularity')
        
        self.glare_thresh = config.getint('DEFAULT', 'glare_thresh')
        
        self._load_calibration(config.get('DEFAULT', 'calibration_file'))
        
        self.smoothing_window = 5
        self.x_history = []
        self.y_history = []
        self.distance_history = []
        self.angle_history = []
        
        if self.show_display:
            self._init_trackbars()
        
        if self.record_video:
            self.video_writer = cv2.VideoWriter('/usr/src/ai/record/detection.avi', 
                                                cv2.VideoWriter_fourcc(*'XVID'), 
                                                8, (self.width, self.height))
        else:
            self.video_writer = None

        self.frame_count = 0
        self.start_time = time.time()
        self.fps = 0.0
        
        self.roi_top = 0
        self.roi_bottom = 320
        self.roi_height = self.roi_bottom - self.roi_top
        
        self.gray_buf = np.zeros((self.roi_height, self.width), dtype=np.uint8)
        self.hsv_buf = np.zeros((self.roi_height, self.width, 3), dtype=np.uint8)
        
        self.kalman = cv2.KalmanFilter(6, 3)
        self.kalman.transitionMatrix = np.array([
            [1, 0, 1, 0, 0, 0],
            [0, 1, 0, 1, 0, 0],
            [0, 0, 1, 0, 0, 0],
            [0, 0, 0, 1, 0, 0],
            [0, 0, 0, 0, 1, 1],
            [0, 0, 0, 0, 0, 1]
        ], np.float32)
        self.kalman.measurementMatrix = np.array([
            [1, 0, 0, 0, 0, 0],
            [0, 1, 0, 0, 0, 0],
            [0, 0, 0, 0, 1, 0]
        ], np.float32)
        self.kalman.processNoiseCov = np.eye(6, dtype=np.float32) * 0.01
        self.kalman.measurementNoiseCov = np.eye(3, dtype=np.float32) * 0.1
        self.kalman.errorCovPost = np.eye(6, dtype=np.float32) * 100
        
        self.last_prediction = None
        self.miss_count = 0
        self.last_valid_result = (0, 0, 0, 0, 0, None)
        self.locked_ball = None  # 锁定最近的球
        self.locked_time = None  # 锁定时间
        self.lock_timeout = 10.0  # 锁定超时时间（秒）

    def _load_calibration(self, calibration_file):
        if calibration_file:
            fs = cv2.FileStorage(calibration_file, cv2.FILE_STORAGE_READ)
            if fs.isOpened():
                self.camera_matrix = fs.getNode('camera_matrix').mat()
                self.distortion_coefficients = fs.getNode('distortion_coefficients').mat()
                fs.release()
            else:
                print(f"无法打开标定文件: {calibration_file}")
                self._set_default_calibration()
        else:
            self._set_default_calibration()

    def _set_default_calibration(self):
        self.camera_matrix = np.array([[self.focal_length, 0, self.width // 2],
                                       [0, self.focal_length, self.height // 2],
                                       [0, 0, 1]], dtype=np.float32)
        self.distortion_coefficients = np.zeros(5, dtype=np.float32)

    def _init_trackbars(self):
        cv2.namedWindow("Settings", cv2.WINDOW_NORMAL)
        cv2.resizeWindow("Settings", 800, 600)
        
        cv2.createTrackbar("Orange H1", "Settings", self.orange_lower[0], 179, lambda x: x)
        cv2.createTrackbar("Orange H2", "Settings", self.orange_upper[0], 179, lambda x: x)
        cv2.createTrackbar("Orange S1", "Settings", self.orange_lower[1], 255, lambda x: x)
        cv2.createTrackbar("Orange S2", "Settings", self.orange_upper[1], 255, lambda x: x)
        cv2.createTrackbar("Orange V1", "Settings", self.orange_lower[2], 255, lambda x: x)
        cv2.createTrackbar("Orange V2", "Settings", self.orange_upper[2], 255, lambda x: x)
        cv2.createTrackbar("O_Open", "Settings", self.orange_iter_open, 5, lambda x: x)
        cv2.createTrackbar("O_Close", "Settings", self.orange_iter_close, 5, lambda x: x)
        
        cv2.createTrackbar("W_Lower", "Settings", self.white_lower, 255, lambda x: x)
        cv2.createTrackbar("W_Upper", "Settings", self.white_upper, 255, lambda x: x)
        cv2.createTrackbar("W_Blur", "Settings", self.white_blur, 15, lambda x: x if x % 2 == 1 else x+1)
        cv2.createTrackbar("W_Open", "Settings", self.white_iter_open, 5, lambda x: x)
        cv2.createTrackbar("W_Close", "Settings", self.white_iter_close, 5, lambda x: x)
        
        cv2.createTrackbar("MinRadius", "Settings", self.min_radius, 50, lambda x: x)
        cv2.createTrackbar("MaxRadius", "Settings", self.max_radius, 200, lambda x: x)
        cv2.createTrackbar("Circularity", "Settings", int(self.min_circularity * 100), 100, lambda x: x)
        cv2.createTrackbar("GlareThresh", "Settings", self.glare_thresh, 255, lambda x: x)

    def update_settings(self):
        self.orange_lower = np.array([
            cv2.getTrackbarPos("Orange H1", "Settings"),
            cv2.getTrackbarPos("Orange S1", "Settings"),
            cv2.getTrackbarPos("Orange V1", "Settings")
        ])
        self.orange_upper = np.array([
            cv2.getTrackbarPos("Orange H2", "Settings"),
            cv2.getTrackbarPos("Orange S2", "Settings"),
            cv2.getTrackbarPos("Orange V2", "Settings")
        ])
        self.orange_iter_open = cv2.getTrackbarPos("O_Open", "Settings")
        self.orange_iter_close = cv2.getTrackbarPos("O_Close", "Settings")
        
        self.white_lower = cv2.getTrackbarPos("W_Lower", "Settings")
        self.white_upper = cv2.getTrackbarPos("W_Upper", "Settings")
        self.white_blur = cv2.getTrackbarPos("W_Blur", "Settings")
        self.white_iter_open = cv2.getTrackbarPos("W_Open", "Settings")
        self.white_iter_close = cv2.getTrackbarPos("W_Close", "Settings")
        
        self.min_radius = cv2.getTrackbarPos("MinRadius", "Settings")
        self.max_radius = cv2.getTrackbarPos("MaxRadius", "Settings")
        self.min_circularity = cv2.getTrackbarPos("Circularity", "Settings") / 100.0
        self.glare_thresh = cv2.getTrackbarPos("GlareThresh", "Settings")

    def _process_white(self, frame):
        self.gray_buf[:] = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        blurred = cv2.GaussianBlur(self.gray_buf, (self.white_blur, self.white_blur), 0)
        _, mask = cv2.threshold(blurred, self.white_lower, self.white_upper, cv2.THRESH_BINARY)
        mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, self.kernel, iterations=self.white_iter_open)
        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, self.kernel, iterations=self.white_iter_close)
        return mask

    def _process_orange(self, frame):
        self.hsv_buf[:] = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
        self.hsv_buf[:, :, 2] = self.gray_buf
        frame_modified = cv2.cvtColor(self.hsv_buf, cv2.COLOR_HSV2BGR)
        mask = cv2.inRange(self.hsv_buf, self.orange_lower, self.orange_upper)
        mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, self.kernel, iterations=self.orange_iter_open)
        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, self.kernel, iterations=self.orange_iter_close)
        return mask

    def _find_contours(self, mask):
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        valid_balls = []
        for contour in contours:
            area = cv2.contourArea(contour)
            if area < 50:
                continue
            (x, y), radius = cv2.minEnclosingCircle(contour)
            if self.min_radius <= radius <= self.max_radius:
                valid_balls.append((x, y, radius * 2))
        return valid_balls

    def detect_balls(self, frame):
        try:
            roi_frame = frame[self.roi_top:self.roi_bottom, 0:self.width]
            undistorted = cv2.undistort(roi_frame, self.camera_matrix, self.distortion_coefficients)
            orange_mask = self._process_orange(undistorted)
            white_mask = self._process_white(undistorted)
            
            orange_balls = self._find_contours(orange_mask)
            white_balls = self._find_contours(white_mask)
            
            all_balls = []
            for x, y, diameter in orange_balls:
                y += self.roi_top
                all_balls.append((x, y, diameter, 1))
            for x, y, diameter in white_balls:
                y += self.roi_top
                all_balls.append((x, y, diameter, 0))
            
            if not all_balls:
                self.locked_ball = None  # 未检测到球时解锁
                self.locked_time = None
                return 0, 0, 0, 0, 0, frame
            
            # 计算每个球的距离
            balls_with_distance = []
            for x, y, diameter, ball_type in all_balls:
                distance = (self.real_diameter * self.focal_length) / max(diameter, 1e-6)
                balls_with_distance.append((x, y, diameter, ball_type, distance))
            
            # 如果已锁定球，尝试匹配最近的球
            if self.locked_ball is not None and time.time() - self.locked_time < self.lock_timeout:
                locked_x, locked_y, locked_diameter, locked_type, locked_distance = self.locked_ball
                # 查找与锁定球最接近的球（基于位置和直径）
                closest_ball = min(balls_with_distance, key=lambda b: (
                    (b[0] - locked_x) ** 2 + (b[1] - locked_y) ** 2 + ((b[2] - locked_diameter) / 10) ** 2
                ), default=None)
                if closest_ball and abs(closest_ball[4] - locked_distance) < 0.2:  # 距离变化小于 0.2m
                    x, y, diameter, ball_type, distance = closest_ball
                    result = self._calculate_metrics((x, y, diameter), frame, ball_type)
                    self.last_valid_result = result
                    self.locked_ball = (x, y, diameter, ball_type, distance)  # 更新锁定球
                    return result
            
            # 未锁定或锁定超时，选择距离最近的球
            closest_ball = min(balls_with_distance, key=lambda b: b[4])  # 按距离排序
            x, y, diameter, ball_type, distance = closest_ball
            result = self._calculate_metrics((x, y, diameter), frame, ball_type)
            self.last_valid_result = result
            self.locked_ball = (x, y, diameter, ball_type, distance)  # 锁定最近的球
            self.locked_time = time.time()  # 记录锁定时间
            return result
        except Exception as e:
            print(f"检测异常: {str(e)}")
            self.locked_ball = None
            self.locked_time = None
            return 0, 0, 0, 0, 0, frame

    def _glare_suppression(self, frame):
        avg_brightness = np.mean(self.gray_buf)
        if avg_brightness < 100:
            return frame
        _, glare_mask = cv2.threshold(self.gray_buf, self.glare_thresh, 255, cv2.THRESH_BINARY)
        return cv2.inpaint(frame, glare_mask, 3, cv2.INPAINT_TELEA)

    def _calculate_metrics(self, ball_info, frame, ball_type=1):
        try:
            x, y, diameter = ball_info
            measurement = np.array([[x], [y], [diameter]], dtype=np.float32)
            
            _ = self.kalman.predict()
            
            if not (0 <= x < self.width and 0 <= y < self.height and diameter > 0):
                self.miss_count += 1
                if self.miss_count > 5:
                    self.kalman.errorCovPost = np.eye(6, dtype=np.float32) * 100
                    self.last_prediction = None
                    return 0, 0, 0, 0, 0, frame
            else:
                self.miss_count = 0
                self.kalman.correct(measurement)
            
            state = self.kalman.statePost
            smooth_x = state[0, 0]
            smooth_y = state[1, 0]
            smooth_diameter = state[4, 0]
            smooth_distance = (self.real_diameter * self.focal_length) / max(smooth_diameter, 1e-6)
            
            if smooth_distance < 0.20:
                self.x_history.clear()
                self.y_history.clear()
                self.distance_history.clear()
                self.angle_history.clear()
                return 0, 0, 0, 0, 0, frame
            
            dx = smooth_x - self.width // 2
            smooth_angle = math.degrees(math.atan(dx / self.focal_length))
            
            self.x_history.append(smooth_x)
            self.y_history.append(smooth_y)
            self.distance_history.append(smooth_distance)
            self.angle_history.append(smooth_angle)
            
            if len(self.x_history) > self.smoothing_window:
                self.x_history.pop(0)
                self.y_history.pop(0)
                self.distance_history.pop(0)
                self.angle_history.pop(0)
            
            smooth_x = sum(self.x_history) / len(self.x_history)
            smooth_y = sum(self.y_history) / len(self.y_history)
            smooth_distance = sum(self.distance_history) / len(self.distance_history)
            smooth_angle = sum(self.angle_history) / len(self.angle_history)
            
            self.last_prediction = (smooth_x, smooth_y, smooth_distance, smooth_angle)
            
            if self.show_display or self.record_video:
                self._draw_debug_info(frame, int(smooth_x), int(smooth_y), smooth_distance, smooth_angle)
            
            return ball_type, smooth_x, smooth_y, smooth_distance, smooth_angle, frame
        except Exception as e:
            print("指标计算异常:", e)
            return 0, 0, 0, 0, 0, frame

    def _draw_debug_info(self, frame, x, y, distance, angle):
        center_x = self.width // 2
        center_y = self.height // 2
        cv2.line(frame, (center_x, 0), (center_x, self.height), (100, 100, 100), 1)
        cv2.line(frame, (0, center_y), (self.width, center_y), (100, 100, 100), 1)
        cv2.circle(frame, (x, y), 10, (0, 255, 0), 2)
        cv2.line(frame, (center_x, center_y), (x, y), (0, 0, 255), 2)
        cv2.putText(frame, f"Dist: {distance:.2f}m", (10, 30),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2)
        cv2.putText(frame, f"Angle: {angle:.1f} deg", (10, 70),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2)
        cv2.putText(frame, f"FPS: {self.fps:.2f}", (10, 110),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2)
        cv2.line(frame, (0, self.roi_bottom), (self.width, self.roi_bottom), (255, 0, 0), 2)

    def detect_and_display(self):
        self.last_frame = None
        self.last_update_time = time.time()
        try:
            while True:
                start_time = time.time()
                ret, frame = self.cap.read()
                if not ret:
                    print("摄像头读取失败")
                    break
                
                if self.last_frame is not None and np.array_equal(frame, self.last_frame):
                    print("[WARN] Frame frozen, skipping detection")
                    time.sleep(0.01)
                    continue
                self.last_frame = frame.copy()
                
                if self.show_display:
                    self.update_settings()
                    orange_mask_disp = self._process_orange(frame[self.roi_top:self.roi_bottom, 0:self.width])
                    white_mask_disp = self._process_white(frame[self.roi_top:self.roi_bottom, 0:self.width])
                    cv2.imshow("Orange Mask", orange_mask_disp)
                    cv2.imshow("White Mask", white_mask_disp)
                
                status, x, y, distance, angle, processed_frame = self.detect_balls(frame)
                
                self.frame_count += 1
                elapsed_time = time.time() - self.start_time
                self.fps = self.frame_count / elapsed_time if elapsed_time > 0 else 0.0
                
                if status:
                    print(f"位置: ({x:.2f}, {y:.2f}), 距离: {distance:.2f}米, 角度: {angle:.1f}度, FPS: {self.fps:.2f}")
                    self.last_update_time = time.time()
                else:
                    if time.time() - self.last_update_time > 1.0:
                        print("[INFO] No ball detected for 1s, resetting")
                        self.last_valid_result = (0, 0, 0, 0, 0, frame)
                
                if self.show_display and processed_frame is not None and processed_frame.size > 0:
                    cv2.imshow("Frame", processed_frame)
                    cv2.resizeWindow("Frame", self.width, self.height)
                
                if self.record_video and self.video_writer is not None and processed_frame is not None and processed_frame.size > 0:
                    self.video_writer.write(processed_frame)
                
                if cv2.waitKey(1) & 0xFF == ord('q'):
                    break
        finally:
            self.cap.release()
            if self.video_writer is not None:
                self.video_writer.release()
            cv2.destroyAllWindows()

if __name__ == "__main__":
    detector = BallDetector(show_display=False, record_video=False)
    detector.detect_and_display()